lib/galaxy/tools/evaluation.py

```diff
@@ -575,8 +574,8 @@ def __write_workdir_file(self, config_filename, content, context, is_template=Tr
             value = fill_template(content, context=context)
         else:
             value = content
-        with open(config_filename, "wb") as f:
-            f.write(smart_str(value))
+        with open(config_filename, "w") as f:
```

Comment:
This was my initial attempt, but that broke the unicode test tool on docker. I might have changed the encoding in the image in the meantime though, I'll double check and point jenkins at your branch.

Reply:
yeah, this still fails in the docker image:
```
FAIL
( unicode_stream ) > Test-3 ... galaxy.tools DEBUG 2018-08-21 07:44:42,132 Validated and populated state for tool request (0.047 ms)
galaxy.tools.actions INFO 2018-08-21 07:44:42,147 Handled output named out_file1 for tool unicode_stream (0.587 ms)
galaxy.tools.actions INFO 2018-08-21 07:44:42,149 Added output datasets to history (1.813 ms)
galaxy.tools.actions INFO 2018-08-21 07:44:42,150 Verified access to datasets for Job[unflushed,tool_id=unicode_stream] (0.026 ms)
galaxy.tools.actions INFO 2018-08-21 07:44:42,151 Setup for job Job[unflushed,tool_id=unicode_stream] complete, ready to flush (1.757 ms)
galaxy.tools.actions INFO 2018-08-21 07:44:42,166 Flushed transaction for job Job[id=3,tool_id=unicode_stream] (15.094 ms)
galaxy.tools.execute DEBUG 2018-08-21 07:44:42,167 Tool [unicode_stream] created job [3] (23.101 ms)
galaxy.tools.execute DEBUG 2018-08-21 07:44:42,171 Executed 1 job(s) for tool unicode_stream request: (38.651 ms)
galaxy.jobs.mapper DEBUG 2018-08-21 07:44:42,516 (3) Mapped job to destination id: local
galaxy.jobs.handler DEBUG 2018-08-21 07:44:42,520 (3) Dispatching to local runner
galaxy.jobs DEBUG 2018-08-21 07:44:42,524 (3) Working directory for job is: /tmp/tmp76q95lht/tmpo0l6a5id/tmpvh6dmpqm/database/job_working_directory_ico6qlt6/000/3
galaxy.jobs DEBUG 2018-08-21 07:44:42,526 (3) Persisting job destination (destination id: local)
galaxy.jobs.runners DEBUG 2018-08-21 07:44:42,534 Job [3] queued (13.951 ms)
galaxy.jobs.handler INFO 2018-08-21 07:44:42,540 (3) Job dispatched
galaxy.jobs.runners ERROR 2018-08-21 07:44:42,582 (3) Failure preparing job
Traceback (most recent call last):
  File "/galaxy/lib/galaxy/jobs/runners/__init__.py", line 187, in prepare_job
    job_wrapper.prepare()
  File "/galaxy/lib/galaxy/jobs/__init__.py", line 874, in prepare
    self.command_line, self.extra_filenames, self.environment_variables = tool_evaluator.build()
  File "/galaxy/lib/galaxy/tools/evaluation.py", line 435, in build
    raise e
  File "/galaxy/lib/galaxy/tools/evaluation.py", line 431, in build
    self.__build_config_files()
  File "/galaxy/lib/galaxy/tools/evaluation.py", line 504, in __build_config_files
    self.__write_workdir_file(config_filename, config_text, param_dict, is_template=is_template)
  File "/galaxy/lib/galaxy/tools/evaluation.py", line 578, in __write_workdir_file
    f.write(value)
UnicodeEncodeError: 'ascii' codec can't encode characters in position 0-5: ordinal not in range(128)
```

Reply:
(https://jenkins.galaxyproject.org/job/docker-framework-py3/38/testReport/)

Reply:
I've added a debug statement and found out that `f.encoding` is `ANSI_X3.4-1968` .

Reply:
Fixed using `io.open(config_filename, "w", encoding='utf-8')` .